### PR TITLE
ENH/API: preserve non-nano in to_datetime

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: pandas/core/tools/datetimes.py [hunk 1]
```diff
@@ -396,12 +400,16 @@ def _convert_listlike_datetimes(
             arg = arg.tz_convert(None).tz_localize("utc")
         return arg
 
-    elif is_datetime64_ns_dtype(arg_dtype):
+    elif is_datetime64_dtype(arg_dtype):
+        arg_dtype = cast(np.dtype, arg_dtype)
+        if not is_supported_unit(get_unit_from_dtype(arg_dtype)):
+            # We go to closest supported reso, i.e. "s"
+            arg = astype_overflowsafe(
+                np.asarray(arg), np.dtype("M8[s]"), coerce=errors == "coerce"
```
Comment (mroeschke) [inline pandas/core/tools/datetimes.py:408]: How is errors="ignore" handled?

Comment (jbrockmendel) [reply pandas/core/tools/datetimes.py:408]: will continue to incorrectly raise for the time being

Comment (mroeschke) [reply pandas/core/tools/datetimes.py:408]: Gotcha. Planning on addressing in this PR or worth adding a TODO/xfail test?

Comment (jbrockmendel) [reply pandas/core/tools/datetimes.py:408]: will add a TODO in my next "assorted" branch if this isn't a show-stopper

Comment (jbrockmendel) [reply pandas/core/tools/datetimes.py:408]: (actually looks like this needs rebase so will add comment here)

File: pandas/_libs/tslibs/np_datetime.pyx [hunk 2]
```diff
@@ -304,6 +304,7 @@ cpdef ndarray astype_overflowsafe(
     cnp.dtype dtype,
     bint copy=True,
     bint round_ok=True,
+    bint coerce=False,
```
Comment (MarcoGorelli) [inline pandas/_libs/tslibs/np_datetime.pyx:307]: perhaps `is_coerce`, for consistency with `tslib`?

File: pandas/core/reshape/tile.py [hunk 3]
```diff
@@ -527,7 +527,12 @@ def _convert_bin_to_numeric_type(bins, dtype):
             raise ValueError("bins must be of timedelta64 dtype")
     elif is_datetime64_dtype(dtype) or is_datetime64tz_dtype(dtype):
         if bins_dtype in ["datetime", "datetime64"]:
-            bins = to_datetime(bins).view(np.int64)
+            bins = to_datetime(bins)
+            if is_datetime64_dtype(bins):
+                # As of 2.0, to_datetime may give non-nano, so we need to convert
+                #  here until the rest of this file recognizes non-nano
+                bins = bins.astype("datetime64[ns]", copy=False)
```
Comment (MarcoGorelli) [inline pandas/core/reshape/tile.py:534]: is there any test that would hit the `else` branch? I think now `is_datetime64_dtype(bins)` is always `True` in the test suite

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
nice! just got a question about coverage

File: pandas/core/tools/datetimes.py [hunk 4]
```diff
@@ -385,6 +387,8 @@ def _convert_listlike_datetimes(
     """
     if isinstance(arg, (list, tuple)):
         arg = np.array(arg, dtype="O")
+    elif isinstance(arg, PandasArray):
+        arg = np.array(arg)
```
Comment (MarcoGorelli) [inline pandas/core/tools/datetimes.py:391]: Which test hits this?

I ran
```
pytest pandas/tests/tools/test_to_datetime.py \
pandas/tests/io/parser/test_parse_dates.py \
pandas/tests/io/parser/usecols/test_parse_dates.py \
pandas/tests/indexes/datetimes/test_constructors.py \
pandas/tests/indexes/datetimes/test_scalar_compat.py \
pandas/tests/series/test_constructors.py \
pandas/tests/tslibs
```
and couldn't hit it

File: pandas/tests/tools/test_to_datetime.py [hunk 5]
```diff
@@ -858,22 +858,31 @@ def test_to_datetime_dt64s_and_str(self, arg, format):
         "dt", [np.datetime64("1000-01-01"), np.datetime64("5000-01-02")]
     )
     def test_to_datetime_dt64s_out_of_bounds(self, cache, dt):
```
Comment (MarcoGorelli) [inline pandas/tests/tools/test_to_datetime.py:860]: should the name of the test change, now that they're no longer out-of-bounds?

File: pandas/tests/tools/test_to_datetime.py [hunk 6]
```diff
@@ -858,22 +858,31 @@ def test_to_datetime_dt64s_and_str(self, arg, format):
...
+        # We cast to the nearest supported reso, i.e. "s"
+        ts = to_datetime(dt, errors="raise", cache=cache)
+        assert isinstance(ts, Timestamp)
+        assert ts.unit == "s"
+        assert ts.asm8 == dt
+
+        ts = to_datetime(dt, errors="coerce", cache=cache)
+        assert isinstance(ts, Timestamp)
+        assert ts.unit == "s"
+        assert ts.asm8 == dt
```
Comment (MarcoGorelli) [inline pandas/tests/tools/test_to_datetime.py:870]: why is checking with both `raise` and `coerce` necessary, given that no errors are thrown in this test anyway? `errors='ignore'` works too

could we either parametrise over that too, or just remove `errors=`?

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]

File: pandas/core/tools/datetimes.py (hunk 4 above)
Comment (jbrockmendel) [reply pandas/core/tools/datetimes.py:391]: if i disable this line we fail `pandas/tests/arrays/test_datetimelike.py::test_from_pandas_array[M8[ns]] - TypeError: Cannot convert PandasDtype to numpy.dtype`

File: pandas/core/reshape/tile.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/core/reshape/tile.py:534]: no extant test, but im not sure its impossible to reach

Comment (MarcoGorelli) [reply pandas/core/reshape/tile.py:534]: perhaps raise an AssertionError in the `else` block? the it won't show up as uncovered

Comment (jbrockmendel) [reply pandas/core/reshape/tile.py:534]: im not concerned about it showing as uncovered, and don't want to break potentially-working code in the wild.

will need to revisit this eventually to handle non-nano without the cast anyway

File: pandas/tests/tools/test_to_datetime.py (hunk 6 above)
Comment (MarcoGorelli) [reply pandas/tests/tools/test_to_datetime.py:1038]: gentle reminder

File: pandas/tests/tools/test_to_datetime.py (hunk 5 above)
Comment (MarcoGorelli) [reply pandas/tests/tools/test_to_datetime.py:1028]: gentle reminder

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
the code changes are fine, I'd just left a couple of comments (which may have got lost in the conversation), so I've just bumped them

File: pandas/_libs/tslibs/np_datetime.pyx (hunk 2 above)
Comment (jbrockmendel) [reply pandas/_libs/tslibs/np_datetime.pyx:311]: sure

File: pandas/tests/tools/test_to_datetime.py (hunk 5 above)
Comment (MarcoGorelli) [reply pandas/tests/tools/test_to_datetime.py]: @jbrockmendel last one, sorry 😄

Comment (jbrockmendel) [reply pandas/tests/tools/test_to_datetime.py]: already renamed to reflect it is only out of ns bounds

**Review by MarcoGorelli — APPROVED** [review]
Looks good to me pending green

File: pandas/tests/tools/test_to_datetime.py (hunk 5 above)
Comment (MarcoGorelli) [reply pandas/tests/tools/test_to_datetime.py]: I missed that, thanks!

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @jbrockmendel

**Pull request merged**